### PR TITLE
Inductive miner

### PR DESCRIPTION
**Pull request by Mona144**
Group 2 implementation of the inductive miner algorithm

### CODE REVIEW

File: practical/ProcessMining/group2/inductive_miner/util/output/pm4py_process_trees.txt
Comment (TJEWH) [inline practical/ProcessMining/group2/inductive_miner/util/output/pm4py_process_trees.txt:1]: I wouldn't expect test output in a utils directory, but that's just a minor remark

File: practical/ProcessMining/group2/inductive_miner/util/logs_with_pm4py.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+import os
+
+from pm4py.objects.log.obj import EventLog, Trace, Event
+from pm4py.algo.discovery.inductive import algorithm as inductive_miner
+import practical.ProcessMining.group2.inductive_miner.src.algo as algo
```
Comment (TJEWH) [inline practical/ProcessMining/group2/inductive_miner/util/logs_with_pm4py.py:5]: you could use a relative import by using packages => `../../src.algo as algo`

File: practical/ProcessMining/group2/inductive_miner/util/logs_with_pm4py.py
Comment (TJEWH) [inline practical/ProcessMining/group2/inductive_miner/util/logs_with_pm4py.py:1]: Everything clear and simple

File: practical/ProcessMining/group2/inductive_miner/src/graph_utils.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+        :return: List of unreachable pairs.
+        """
+        reach = self.all_pairs_reachability_dag()
+        nodes = self.get_all_nodes()
+        non_reachable_pairs = set()
+        
+        for i in range(len(nodes)):
+            for j in range(i + 1, len(nodes)):
+                u = nodes[i]
+                v = nodes[j]
```
Comment (TJEWH) [inline practical/ProcessMining/group2/inductive_miner/src/graph_utils.py:245]: Would be more straight forward to use enumerate() for that purpose instead of range(len())

File: practical/ProcessMining/group2/inductive_miner/src/dfg_visualisation.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+
+
+
+#TODO add additional event logs and tests
+
+
+
+l1 = EventLog({'abcd': 3,   'acbd' : 4, 'abcefbcd': 2,  'abcefcbd' : 1, 'acbefbcd' : 2, 'acbefbcefcbd' : 1})
+l2 = EventLog({'bc': 3,   'cb' : 4, 'bcefbc': 2,  'bcefcb' : 1, 'cbefbc' : 2, 'cbefbcefcb' : 1})
+
```
Comment (TJEWH) [inline practical/ProcessMining/group2/inductive_miner/src/dfg_visualisation.py:13]: Why do we need hard coded events?

File: practical/ProcessMining/group2/inductive_miner/src/dfg_visualisation.py
Comment (TJEWH) [inline practical/ProcessMining/group2/inductive_miner/src/dfg_visualisation.py:1]: Argueable, if this code snippet needs it's own file. If yes, I would refactor everything to functions which get properly called in __main__

File: practical/ProcessMining/group2/inductive_miner/tests/test_graph_utils.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        'e': ['f'],
+        'f': [],
+        'g': [],
+    }
+
+
+def test_add_edge():
+    # Test if edge is added correctly
+    g = Graph({'a': ['b'], 'b': ['c', 'd'], 'c': [], 'd': []})
+    g.add_edge('a', 'c')
```
Comment (TJEWH) [inline practical/ProcessMining/group2/inductive_miner/tests/test_graph_utils.py:31]: how is it handling false edges e.g.?

File: practical/ProcessMining/group2/inductive_miner/tests/test_graph_utils.py
Comment (TJEWH) [inline practical/ProcessMining/group2/inductive_miner/tests/test_graph_utils.py:1]: Test qualitity could be improved by using a testing framework to avoid duplicate code. Also for some test more edge case testing could make sense

File: practical/ProcessMining/group2/inductive_miner/tests/test_algo.py
Comment (TJEWH) [inline practical/ProcessMining/group2/inductive_miner/tests/test_algo.py:1]: Also here a test quality could be improved by a certain testing framework for parametrization of tests, most tested cases appear to be trivial

File: practical/ProcessMining/group2/inductive_miner/src/algo.py [hunk 5]
```diff
@@ -0,0 +1,527 @@
...
+            traces = {}
+            for line in file:
+                trace = line.strip()
+                if trace in traces:
+                    traces[trace] += 1
+                else:
+                    traces[trace] = 1
+            return EventLog(traces)
+
+class DirectlyFollowsGraph(Graph):
```
Comment (TJEWH) [inline practical/ProcessMining/group2/inductive_miner/src/algo.py:35]: I like that you are introducing seperate classes for seperate parts, but in my eyes logic related step would be to move the parts to seperate files for strict seperation of concerns.

File: practical/ProcessMining/group2/inductive_miner/src/algo.py [hunk 6]
```diff
@@ -0,0 +1,527 @@
...
+        :return: EventLog object.
+        """
+        with open(os.path.join(script_dir, file_path), 'r') as file:
+            traces = {}
+            for line in file:
+                trace = line.strip()
+                if trace in traces:
+                    traces[trace] += 1
+                else:
+                    traces[trace] = 1
```
Comment (TJEWH) [inline practical/ProcessMining/group2/inductive_miner/src/algo.py:32]: using defaultdict(int) instead

File: practical/ProcessMining/group2/inductive_miner/src/algo.py [hunk 7]
```diff
@@ -0,0 +1,527 @@
...
+        """
+        # Convert the graph to undirected
+        undirected = dfg.convert_to_undirected()
+        # Find connected components
+        cuts = undirected.find_components()
+
+        return None if len(cuts) == 1 else cuts
+    
+    def find_sequence_cut(self, dfg: DirectlyFollowsGraph) -> List[List[str]]:
+        """
```
Comment (TJEWH) [inline practical/ProcessMining/group2/inductive_miner/src/algo.py:115]: Sequence cuts logic is nasty, but refactoring it into seperate methods for each step could improve readability

**Review by TJEWH — COMMENTED** [review]
All in all a nice solution with way better code quality than in task 2. 

Main point of critisim is the project structure, refactoring algo.py in seperate files and renaming the other files in src, project structure would become more comprehensive

File: practical/ProcessMining/group2/inductive_miner/src/dfg_visualisation.py (hunk 3 above)
Comment (reichertli) [reply practical/ProcessMining/group2/inductive_miner/src/dfg_visualisation.py:13]: This was just for testing purposes, it will be removed in further refactoring

**Review by OhItsLena — COMMENTED** [review]
Nice implementation with smart ideas :) I shows that having a class for Graph, EventLog, etc. pays of with regards to readability. Tests could be improved and maybe some more development comments for complex logic like the cuts and splits.

File: practical/ProcessMining/group2/inductive_miner/src/algo.py (hunk 7 above)
Comment (Mona144) [reply practical/ProcessMining/group2/inductive_miner/src/algo.py:115]: Agree, but abandoned due to time and sanity reasons

File: practical/ProcessMining/group2/inductive_miner/src/algo.py [hunk 8]
```diff
@@ -0,0 +1,527 @@
...
+                    if node not in self.graph: # Isolated nodes
+                        self.graph[node] = []
+            else:
+                self.graph[''] = [] # Add tau node for empty traces
+
+    def __str__(self) -> str:
+        """Represent Directly Follows Graph as string."""
+        return f"Directly Follows Graph: (\n\tGraph: {dict(self.graph)}\n\tStart nodes: {self.start_nodes}\n\tEnd nodes: {self.end_nodes}\n)"
+
+class ProcessTree:
```
[the remark this answers is not in the document]
Comment (Mona144) [reply practical/ProcessMining/group2/inductive_miner/src/algo.py:73]: Agreed, tried to apply this refactoring but some tests broke in a non-tivial manner, so for the sake of sanity and peace we will leave this as is
